Deprecate `ByteSourceRange` in favor of `Range<AbsolutePosition>` (#2588)
apple/swift-syntax: Release Notes/600.md

  - Pull request: https://github.com/apple/swift-syntax/pull/2594

- `SyntaxIdentifier` conformance to `Comparable`:
  - Description: A `SyntaxIdentifier` compares less than another `SyntaxIdentifier` if the node at that identifier occurs first during a depth-first traversal of the tree.
  - Pull request: https://github.com/apple/swift-syntax/pull/2594

- `SyntaxIdentifier.indexInTree` and `SyntaxIdentifier.fromIndexInTree`
  - Description: `SyntaxIdentifier.indexInTree` allows the retrieval of a `SyntaxIdentifier` that identifies the syntax node independent of the syntax tree. `SyntaxIdentifier.fromIndexInTree` allows the creation for a `SyntaxIdentifier` from a tree-agnostic `SyntaxIdentifier.IndexInTree` and the tree's root node.
  - Pull request: https://github.com/apple/swift-syntax/pull/2594

- `Range<AbsolutePosition>`

>   - Description: `Range<AbsolutePosition>` gained a few convenience functions inspired from `ByteSourceRange`: `init(position:length:)`,  `length`, `intersectsOrTouches`, `intersects`, `intersecting`

Reviewer: Could you update this for `overlaps()`/`clamped(to:)` style APIs?

Author: Thanks for noticing. I forgot about the release notes

  - Description: With the change to parse `#if canImport(MyModule, _version: 1.2.3)` as a function call instead of a dedicated syntax node, `1.2.3` natively gets parsed as a member access `3` to the `1.2` float literal. This property allows the reinterpretation of such an expression as a version tuple.
  - Pull request: https://github.com/apple/swift-syntax/pull/2025

- `Range<AbsolutePosition>`

>   - Description: `Range<AbsolutePosition>` gained a few convenience functions inspirec from `ByteSourceRange`: `init(position:length:)`,  `length`, `intersectsOrTouches`, `intersects`, `intersecting`

Reviewer 2: ```suggestion
  - Description: `Range<AbsolutePosition>` gained a few convenience functions inspired from `ByteSourceRange`: `init(position:length:)`,  `length`, `intersectsOrTouches`, `intersects`, `intersecting`
```

  - Description: Instead of parsing `canImport` inside `#if` directives as a special expression node, parse it as a functionc call expression. This is in-line with how the `swift(>=6.0)` and `compiler(>=6.0)` directives are parsed.
  - Pull request: https://github.com/apple/swift-syntax/pull/2025

- `SyntaxClassifiedRange.offset`, `length` and `endOffset`

>   - Description: Instead of providing these properties work with the replaced `range` instead.

Reviewer 2: ```suggestion
  - Description: Deprecated these properties in favor of `range`.
```

  - Pull request: https://github.com/apple/swift-syntax/pull/2587

- `SyntaxProtocol.totalByteRange` and `trimmedByteRange`

>   - Description: Renamed to `range` and `trimmedRange` with the deprecation of `ByteSourceRange` in favor of `Range<AbsolutePosition>`

Reviewer 2: ```suggestion
  - Description: Renamed to `range` and `trimmedRange`, both now returning a `Range<AbsolutePosition>`.
```

  - Pull request: https://github.com/apple/swift-syntax/pull/2587

- `ByteSourceRange` deprecated in favor of `Range<AbsolutePosition>`

>   - Description: Instead of having a dedicated range type in SwiftSyntax, use `Range<AbsolutePosition>` instead. `Range<AbsolutePosition>` has deprecated compatibility layers to make it API-compatible with `ByteSourceRange`

Reviewer 2: ```suggestion
  - Description: `ByteSourceRange` is being dropped for `Range<AbsolutePosition>`, where the latter clearly signifies that it uses UTF-8 byte positions. `Range<AbsolutePosition>` has deprecated compatibility layers to make it API-compatible with `ByteSourceRange`
```
